internal/controller/ledger/controller_default.go

```diff
@@ -195,71 +195,78 @@ func (ctrl *DefaultController) Import(ctx context.Context, stream chan ledger.Lo
-				if _, _, err := ctrl.store.DeleteTransactionMetadata(ctx, payload.TargetID.(int), payload.Key, log.Date); err != nil {
-					return nil, fmt.Errorf("failed to delete transaction metadata: %w", err)
+			case ledger.SavedMetadata:
+				switch payload.TargetType {
+				case ledger.MetaTargetTypeTransaction:
+					logging.FromContext(ctx).Debugf("Saving metadata of transaction %d", payload.TargetID)
+					if _, _, err := ctrl.store.UpdateTransactionMetadata(ctx, payload.TargetID.(int), payload.Metadata, log.Date); err != nil {
+						return nil, fmt.Errorf("failed to update transaction metadata: %w", err)
+					}
+				case ledger.MetaTargetTypeAccount:
+					logging.FromContext(ctx).Debugf("Saving metadata of account %s", payload.TargetID)
+					if err := ctrl.store.UpdateAccountsMetadata(ctx, ledger.AccountMetadata{
+						payload.TargetID.(string): payload.Metadata,
+					}); err != nil {
+						return nil, fmt.Errorf("failed to update account metadata: %w", err)
+					}
 				}
-			case ledger.MetaTargetTypeAccount:
-				logging.FromContext(ctx).Debugf("Deleting metadata of account %s", payload.TargetID)
-				if err := ctrl.store.DeleteAccountMetadata(ctx, payload.TargetID.(string), payload.Key); err != nil {
-					return nil, fmt.Errorf("failed to delete account metadata: %w", err)
+			case ledger.DeletedMetadata:
+				switch payload.TargetType {
+				case ledger.MetaTargetTypeTransaction:
+					logging.FromContext(ctx).Debugf("Deleting metadata of transaction %d", payload.TargetID)
+					if _, _, err := ctrl.store.DeleteTransactionMetadata(ctx, payload.TargetID.(int), payload.Key, log.Date); err != nil {
+						return nil, fmt.Errorf("failed to delete transaction metadata: %w", err)
+					}
+				case ledger.MetaTargetTypeAccount:
+					logging.FromContext(ctx).Debugf("Deleting metadata of account %s", payload.TargetID)
+					if err := ctrl.store.DeleteAccountMetadata(ctx, payload.TargetID.(string), payload.Key); err != nil {
+						return nil, fmt.Errorf("failed to delete account metadata: %w", err)
+					}
 				}
 			}
-		}
 
-		logCopy := log
-		logging.FromContext(ctx).Debugf("Inserting log %d", *log.ID)
-		if err := ctrl.store.InsertLog(ctx, &log); err != nil {
-			return nil, fmt.Errorf("failed to insert log: %w", err)
-		}
+			logCopy := log
+			logging.FromContext(ctx).Debugf("Inserting log %d", *log.ID)
+			if err := ctrl.store.InsertLog(ctx, &log); err != nil {
+				return nil, fmt.Errorf("failed to insert log: %w", err)
+			}
 
-		if ctrl.ledger.HasFeature(features.FeatureHashLogs, "SYNC") {
-			if !reflect.DeepEqual(log.Hash, logCopy.Hash) {
-				return nil, newErrInvalidHash(*log.ID, logCopy.Hash, log.Hash)
+			if ctrl.ledger.HasFeature(features.FeatureHashLogs, "SYNC") {
+				if !reflect.DeepEqual(log.Hash, logCopy.Hash) {
+					return nil, newErrInvalidHash(*log.ID, logCopy.Hash, log.Hash)
+				}
 			}
-		}
 
-		return nil, nil
-	})
+			return nil, nil
+		},
+		trace.WithNewRoot(),
```

Comment:
@laouji Just these lines. Sorry, I should not have added a eol earlier.

Reply:
Thanks.